gh-90815: Add mimalloc memory allocator

DinoV opened the pull request:
<!-- gh-issue-number: gh-90815 -->
* Issue: gh-90815
<!-- /gh-issue-number -->

This adds mimalloc as an optional (but preferred when available) allocator to CPython.  This is a bit of a mashup of the work from https://github.com/python/cpython/pull/109914 and the work of @colesbury to use mimalloc for no-gil and various updates to bring it up to current CPython.

The configuration logic added by @tiran is re-used and we keep pymalloc support unlike in the version from @colesbury.  mimalloc is updated to 2.12 and along with a few changes @colesbury made to it.

This has run into some issues with subinterpreter support in that the allocator's are now stored in thread state and are per-thread.  Sub interpreters in some scenarios will create a thread state on one thread and run that on another thread.  Most of these are documented in the code base as being known issues.  I've modified these so that we will find the right thread based upon the current thread ID and switch to it rather than getting the head thread.  This seems pretty reasonable but looks pretty weird when we need to do it at interpreter shutdown.

colesbury committed:
Add mimalloc v2.0.9

Modified src/alloc.c to remove include of alloc-override.c and not
compile new handler.

Ignore mimalloc symbols for now.

Did not include the following files:

 - include/mimalloc-new-delete.h
 - include/mimalloc-override.h
 - src/alloc-override-osx.c
 - src/alloc-override.c
 - src/static.c
 - src/region.c

colesbury committed:
mimalloc: minimal changes for use in Python

 - remove debug spam for freeing large allocations
 - use same bytes (0xDD) for freed allocations in CPython and mimalloc
   This is important for the test_capi debug memory tests

DinoV committed:
Upgrade to mimalloc 2.12

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
mimalloc: enable as allocator option

DinoV committed:
Blurb

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
Don't export symbol

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
Fix mimalloc formatting

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
Add includes to analyzer

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

corona10 commented on configure.ac:
```diff
@@ -4489,6 +4489,73 @@ then
+  AC_SUBST([MIMALLOC_OBJS], ['$(MIMALLOC_OBJS)'])
+fi
+
+AC_MSG_RESULT([$with_mimalloc])
```
IIUC, `./configure --disable-gil` should enable `--with-mimalloc` by default right?

DinoV replied:
Ultimately it should, but I figure we can add that when a PR that actually integrates the two shows up.

corona10 reviewed (commented):
And also, would you like to add mimalloc CI to GA too?
WDYT @vstinner and @erlend-aasland ?

DinoV committed:
Add mimalloc MIT license

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

vstinner commented on Modules/_testcapi/mem.c:
```diff
@@ -613,5 +613,14 @@ _PyTestCapi_Init_Mem(PyObject *mod)
+#else
+    v = Py_False;
+#endif
+    if (PyModule_AddObjectRef(mod, "WITH_MIMALLOC", v) < 0) {
```
Please log this value in Lib/test/pythoninfo.py, around WITH_PYMALLOC.

Note: pymem_getallocatorsname is already logged as pymem.allocator key, that's fine.

vstinner commented on Include/internal/pycore_pymem_init.h:
```diff
@@ -18,17 +18,31 @@ extern void * _PyMem_RawRealloc(void *, void *, size_t);
+extern void * _PyObject_MiMalloc(void *ctx, size_t nbytes);
+extern void* _PyObject_MiCalloc(void *ctx, size_t nelem, size_t elsize);
+extern void* _PyObject_MiRealloc(void *ctx, void *ptr, size_t nbytes);
+extern void _PyObject_MiFree(void *ctx, void *ptr);
```
Some consistency, please :-)

```suggestion
extern void* _PyMem_MiMalloc(void *ctx, size_t size);
extern void* _PyMem_MiCalloc(void *ctx, size_t nelem, size_t elsize);
extern void* _PyMem_MiRealloc(void *ctx, void *ptr, size_t size);
extern void _PyMem_MiFree(void *ctx, void *ptr);

extern void* _PyObject_MiMalloc(void *ctx, size_t nbytes);
extern void* _PyObject_MiCalloc(void *ctx, size_t nelem, size_t elsize);
extern void* _PyObject_MiRealloc(void *ctx, void *ptr, size_t nbytes);
extern void _PyObject_MiFree(void *ctx, void *ptr);
```

vstinner commented on Misc/NEWS.d/next/Core and Builtins/2023-09-06-12-36-11.bpo-46657.xea1T_.rst:
```diff
@@ -0,0 +1 @@
+[WIP] Add mimalloc memory allocator support.
```
Please remove [WIP].

vstinner reviewed (commented):
It would be nice to add a section about mimalloc in https://docs.python.org/dev/c-api/memory.html

Also, the new init options should be documented at:
https://docs.python.org/dev/c-api/init_config.html#c.PyPreConfig.allocator

New config option should be documented at: https://docs.python.org/dev/using/configure.html

vstinner reviewed (commented):
Should Include/mimalloc/ be installed? Should it be used outside Python?

vstinner commented on Tools/build/smelly.py:
```diff
@@ -7,7 +7,7 @@
 
 
-ALLOWED_PREFIXES = ('Py', '_Py')
+ALLOWED_PREFIXES = ('Py', '_Py', 'mi_', '_mi_')
```
It's kind of unfortunable :-( These symbols should not exported, no?

DinoV replied:
Looks like Christian solved this by using mimalloc's `static.c` and a set of re-defs of the exports to be `_Py` prefixed,  I'll re-use that approach (and prefix with `PyUnstable`)

colesbury replied:
I think `static.c` is for statically overriding the standard library's `malloc()`, `free()`, etc. (and C++'s `new` operator). We don't want to do that.

colesbury replied:
To clarify the above comment, we can probably do the mostly the same thing as `static.c` but don't include the files used for overriding the standard library:

* alloc.c
* alloc-aligned.c
* alloc-posix.c
* prim/osx/alloc-override-zone.c

vstinner replied:
Would it be possible to have a script which imports mimalloc and applies a patch on top of it to rename mi and _mi symbols to add a _Py prefix? Or at least a patch?

We do have a patch on on top of Python/dtoa.c which is maintained Python for example. My notes on vendored libraries: https://pythondev.readthedocs.io/files.html#vendored-external-libraries

DinoV replied:
@vstinner Christian included the command he used in his `pycore_mimalloc.h` and it just produces a bunch of `#define`s to do the renames, and then no patches are necessary to mimalloc.  It seems like a pretty clean solution to me?

@colesbury Ahh, you're right...  I'll just re-drop `static.c` and do the same set of .c `#include`s it does in `obmalloc.c` minus the alloc override ones.  We'll just need to update things with upgrades if the file set changes.

DinoV replied:
@colesbury Oh, actually `static.c` will only override the built-in allocators if `MI_IN_ALLOC_C` is defined!   It looks like `alloc-aligned.c` and `alloc-posix.c` only bring along `mi` prefixed functions, and `alloc-override-zone.c` looks for `MI_MALLOC_OVERRIDE`.    And we do want `alloc.c` as it defines `mi_free`, `mi_malloc` and friends :)

vstinner replied:
Do we have to **export** these mi_alloc() and mi_free() symbols? Can't we make them private, and let C extensions call PyMem_Malloc() and PyMem_Free()?

DinoV replied:
So I don't think anything is actually exported here with the current build.  If you do `./configure --enable-shared` and run `nm libpython3.13.so.1.0` all of the mi_malloc related symbols are of type `t` meaning they are not global.

Likewise if you run `nm ./python` you'll see that in the statically linked `python` that all of the symbols are private as well.

It's only when we run `nm` on the static library directly that they're still listed as being global.  I assume this is some artifact of the fact that this still needs to be linked in and things will get updated after that - but I must admit I don't fully understand this.

I think this is mainly a consequence of Python's build specifying that default visibility is hidden.  mimalloc will specifically mark symbols it wants as exported with `__attribute__((visibility("default")))` when it wants them exported (and on Windows will do `__declspec(dllexport)`).  

The big reason for the additional symbols that smelly is picking up though is that CPython uses `static` very liberally throughout the code base and mimalloc spreads its implementation across multiple files not using `static`. 

I think probably the most straight forward fix here is to allow `_mi` and `mi` prefixes for only static library, and disallow them for the shared library.  This will assure that there are no exported symbols that shouldn't be there but will also keep the good hygine of using `static` where appropriate in the CPython code base.

DinoV committed:
Log mimalloc in Lib/test/pythoninfo.py
Document new mimalloc support
Cleanup some formatting

DinoV committed:
Use macro defs for exports as done in https://github.com/python/cpython/pull/31164/

DinoV committed:
Fix formatting

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
Fix ./python -m test test_sys -R 3:3

mimalloc is thread safe and shares a single heap across all runtimes,
therefore finalization and getting global allocated blocks across
all runtimes is different.

DinoV committed:
Move mimalloc headers to internal, remove re-defs, update smelly
to allow mi malloc symbols in static lib, but not shared lib

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
Fix header locations on Windows
Fix header locations in c-analyzer
Fix smelly trailing white space

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

DinoV committed:
Try again to fix windows headers

DinoV committed:
Make memory sanitizer happy

vstinner commented on Include/internal/pycore_pymem_init.h:
```diff
@@ -18,17 +18,31 @@ extern void * _PyMem_RawRealloc(void *, void *, size_t);
+extern void _PyObject_MiFree(void *ctx, void *ptr);
+
+#  define PYMEM_ALLOC {NULL, _PyMem_MiMalloc, _PyMem_MiCalloc, _PyMem_MiRealloc, _PyMem_MiFree}
+#  define PYOBJ_ALLOC {NULL, _PyObject_MiMalloc, _PyObject_MiCalloc, _PyObject_MiRealloc, _PyObject_MiFree}
```
I would prefer to no use mimalloc by default for now. Not in this PR.

Maybe you can write a follow PR to use it by default if Python is built by NoGIL?

vstinner reviewed (commented):
For me, the change is mostly ok. But I would prefer that *this* PR doesn't switch the default Python memory allocator to mimalloc. Please write a separated PR for that. Also, I would prefer to move slowly, and only use mimalloc by default if Python is built with NoGIL.

Ah nice, the symbols are no longer exported:

```
$ ./configure --enable-shared --with-pydebug && make
$ readelf -s libpython3.13d.so.1.0 |grep mi_malloc$
 17640: 00000000001e2e31    37 FUNC    LOCAL  DEFAULT   12 mi_malloc
```

For comparison, PyLong_AsLong() is exported:

```
$ readelf -s libpython3.13d.so.1.0 |grep PyLong_AsLong$
  1694: 00000000001abc65    76 FUNC    GLOBAL DEFAULT   12 PyLong_AsLong
 18422: 00000000001abc65    76 FUNC    GLOBAL DEFAULT   12 PyLong_AsLong
```

Moreover, mimalloc/ headers moved to Include/internal/: good, good.

Support for mimalloc is always enabled: I think that it's a good think, that what I proposed before.

vstinner commented on Tools/build/smelly.py:
```diff
@@ -11,6 +11,11 @@
+# mimalloc doesn't use static, but it's symbols are not exported
+# from the shared library.  They do show up in the static library
+# before its linked into an executable.
+ALLOWED_STATIC_PREFIXES = ('mi_', '_mi_')
```
Thank you for distinguishing local and exported symbols. It's good that "mi" and "_mi_" prefixes are only allowed for local symbols.

DinoV committed:
Keep pymalloc the default

vstinner commented on Lib/test/pythoninfo.py:
```diff
@@ -543,7 +543,8 @@ def collect_sysconfig(info_add):
         'WITH_DTRACE',
-        'WITH_FREELISTS',
+        'WITH_FREELISTS'
+        'WITH_MIMALLOC',
```
```suggestion
        'WITH_FREELISTS',
        'WITH_MIMALLOC',
```

vstinner committed:
Update Lib/test/pythoninfo.py

vstinner committed:
Fix memory doc

mimalloc is not the default

vstinner committed:
license.rst: remove empty line at the end

vstinner committed:
Reduce Makefile.pre.in indentation

vstinner committed:
obmalloc.c: indent

vstinner committed:
Split _PyObject_DebugMallocStats() in sub-functions

vstinner reviewed (commented):
See also issue #111499: PYTHONMALLOCSTATS=1 fails with fatal error at Python exit. Using mimalloc, Python is not affected :-)
